pkg/tracing/tracing.go

```diff
@@ -83,71 +294,265 @@ func New(spec *Spec) (*Tracer, error) {
 		return NoopTracer, nil
 	}
 
-	endpoint, err := zipkingo.NewEndpoint(spec.ServiceName, spec.Zipkin.Hostport)
-	if err != nil {
-		return nil, err
-	}
+	opts := []sdktrace.TracerProviderOption{
+		sdktrace.WithRawSpanLimits(spec.newSpanLimits()),
+		sdktrace.WithSampler(spec.newSampler())}
 
-	sampler, err := zipkingo.NewBoundarySampler(spec.Zipkin.SampleRate, fasttime.Now().Unix())
-	if err != nil {
-		return nil, err
+	if r, err := spec.newResource(); err == nil {
+		opts = append(opts, sdktrace.WithResource(r))
+	} else {
+		return NoopTracer, err
 	}
 
-	var reporter zipkinreporter.Reporter
-	if spec.Zipkin.DisableReport {
-		reporter = zipkinreporter.NewNoopReporter()
+	if sps, err := spec.newBatchSpanProcessors(); err == nil {
+		for _, sp := range sps {
+			opts = append(opts, sdktrace.WithSpanProcessor(sp))
+		}
 	} else {
-		reporter = zipkingohttp.NewReporter(spec.Zipkin.ServerURL)
-	}
-	tracer, err := zipkingo.NewTracer(
-		reporter,
-		zipkingo.WithLocalEndpoint(endpoint),
-		zipkingo.WithSharedSpans(spec.Zipkin.SameSpan),
-		zipkingo.WithTraceID128Bit(spec.Zipkin.ID128Bit),
-		zipkingo.WithSampler(sampler),
-		zipkingo.WithTags(spec.Tags),
+		return NoopTracer, err
+	}
+
+	tp := sdktrace.NewTracerProvider(
+		opts...,
 	)
-	if err != nil {
-		return nil, err
+
+	return &Tracer{Tracer: tp.Tracer(""), tp: tp, propagator: spec.newPropagator()}, nil
+}
+
+func (spec *Spec) newResource() (*resource.Resource, error) {
+
+	var attrs = []attribute.KeyValue{semconv.ServiceNameKey.String(spec.ServiceName)}
+	for k, v := range spec.Attributes {
+		attrs = append(attrs, attribute.String(k, v))
 	}
 
-	return &Tracer{
-		tracer: tracer,
-		closer: reporter,
-	}, nil
+	return resource.Merge(resource.Default(),
+		resource.NewWithAttributes(semconv.SchemaURL, attrs...))
```

Comment:
per https://github.com/golang/go/wiki/CodeReviewComments#line-length
propose not breaking a lengthy statement into several lines.

if it is too long, try breaking it into several statements, this makes the code more readable.

there are other cases.